File: web/src/services/api/groupBridge.ts
import { apiJson } from "./base";

// Group Bridge remote-send client.

export interface GroupBridgeRegistration {
  registration_id: string;
  group_id: string;
  url: string;
  transport: string;
  remote_group_id: string;
  remote_peer_id?: string;
  credential_ref: string;
  user_id: string;
  status: string;
  created_at: string;
  updated_at: string;
  last_sync_at?: string | null;
  last_error?: string | null;
}

export interface GroupBridgeDeliveryError {
  code: string;
  message: string;
  retriable?: boolean;
  transport?: string;
  http_status?: number | null;
}

export interface GroupBridgeDeliveryReceipt {
  status: string;
  ok?: boolean;
  registration_id?: string;
  idempotency_key?: string;
  remote_event_id?: string | null;
  transport?: string;
  error?: GroupBridgeDeliveryError | null;
}

export interface GroupBridgeIdentity {
  node_id: string;
  peer_id: string;
}

export type GroupBridgeAccessLevel = "messages" | "read" | "full";

export interface GroupBridgePairingInvite {
  invite_id: string;
  group_id: string;
  remote_group_id: string;
  remote_peer_id: string;
  transport: string;
  status: string;
  expires_at: string;
  pairing_code?: string;
  request_id?: string;
}

export interface GroupBridgePairingRequest {
  request_id: string;
  invite_id: string;
  group_id: string;
  remote_group_id: string;
  remote_peer_id: string;
  status: string;
  registration_id?: string;
  approved_by?: string;
  rejected_by?: string;
  rejection_reason?: string;
}

export interface GroupBridgeTrust {
  trust_id: string;
  request_id: string;
  registration_id: string;
  group_id: string;
  remote_group_id: string;
  remote_group_title?: string;
  remote_endpoint?: string;
  remote_peer_id: string;
  transport: string;
  status: string;
  access_level?: GroupBridgeAccessLevel | string;
  remote_access_level?: GroupBridgeAccessLevel | string;
  access_updated_by?: string;
}

export interface GroupBridgePairingOutbound {
  outbound_id: string;
  local_group_id: string;
  issuer_endpoint: string;
  issuer_group_id: string;
  issuer_group_title?: string;
  issuer_peer_id: string;
  invite_id: string;
  status: string;
  remote_request?: Record<string, unknown>;
  last_error?: string;
  created_at?: string;
  updated_at?: string;
}

export interface PairingInviteInput {
  groupId: string;
  remoteGroupId?: string;
  remotePeerId?: string;
  ttlSeconds?: number;
}

export interface PairingRequestInput {
  pairingCode: string;
  requesterGroupId: string;
  requesterPeerId: string;
  inviteId?: string;
}

export interface RemotePairingRequestInput {
  localGroupId: string;
  localGroupTitle?: string;
  payload: Record<string, unknown>;
}

export interface PairingConnectionInfoInput {
  groupId: string;
  inviteId: string;
  issuerEndpoint: string;
  issuerGroupTitle?: string;
}

export async function fetchGroupBridgeStatus(groupId?: string) {
  const suffix = groupId ? `?group_id=${encodeURIComponent(groupId)}` : "";
  return apiJson<{ registrations: GroupBridgeRegistration[] }>(`/api/group-bridge/status${suffix}`);
}

export async function unregisterGroupBridge(registrationId: string) {
  return apiJson<{ deleted: boolean }>("/api/group-bridge/unregister", {
    method: "POST",
    body: JSON.stringify({ registration_id: registrationId }),
  });
}

export async function fetchGroupBridgeDeliveryStatus(registrationId: string, idempotencyKey: string) {
  return apiJson<{ receipt: GroupBridgeDeliveryReceipt | null }>(
    `/api/group-bridge/registrations/${encodeURIComponent(registrationId)}/deliveries/${encodeURIComponent(idempotencyKey)}`,
  );
}

export async function fetchGroupBridgeIdentity() {
  return apiJson<{ identity: GroupBridgeIdentity }>("/api/group-bridge/pairing/identity");
}

export async function createGroupBridgePairingInvite(input: PairingInviteInput) {
  return apiJson<{ invite: GroupBridgePairingInvite }>("/api/group-bridge/pairing/invites", {
    method: "POST",
    body: JSON.stringify({
      group_id: input.groupId,
      remote_group_id: input.remoteGroupId,
      remote_peer_id: input.remotePeerId,
      ttl_seconds: input.ttlSeconds ?? 600,
    }),
  });
}

export async function createGroupBridgePairingConnectionInfo(input: PairingConnectionInfoInput) {
  return apiJson<{ payload: Record<string, unknown> }>("/api/group-bridge/pairing/connection-info", {
    method: "POST",
    body: JSON.stringify({
      group_id: input.groupId,
      invite_id: input.inviteId,
      issuer_endpoint: input.issuerEndpoint,
      issuer_group_title: input.issuerGroupTitle ?? "",
    }),
  });
}

export async function createGroupBridgePairingRequest(input: PairingRequestInput) {
  const body: Record<string, unknown> = {
    pairing_code: input.pairingCode,
    requester_group_id: input.requesterGroupId,
    requester_peer_id: input.requesterPeerId,
  };
  if (input.inviteId) {
    body.invite_id = input.inviteId;
  }
  return apiJson<{ request: GroupBridgePairingRequest }>("/api/group-bridge/pairing/requests", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export async function createGroupBridgeRemotePairingRequest(input: RemotePairingRequestInput) {
  return apiJson<{ outbound: Record<string, unknown> }>("/api/group-bridge/pairing/remote-requests", {
    method: "POST",
    body: JSON.stringify({
      local_group_id: input.localGroupId,
      local_group_title: input.localGroupTitle ?? "",
      payload: input.payload,
    }),
  });
}

export async function fetchGroupBridgePairingRequests(groupId?: string) {
  const suffix = groupId ? `?group_id=${encodeURIComponent(groupId)}` : "";
  return apiJson<{ requests: GroupBridgePairingRequest[] }>(`/api/group-bridge/pairing/requests${suffix}`);
}

export async function approveGroupBridgePairingRequest(requestId: string, approverUserId = "") {
  return apiJson<{ request: GroupBridgePairingRequest; registration: GroupBridgeRegistration; trust: GroupBridgeTrust | null }>(
    `/api/group-bridge/pairing/requests/${encodeURIComponent(requestId)}/approve`,
    {
      method: "POST",
      body: JSON.stringify({ approver_user_id: approverUserId }),
    },
  );
}

export async function rejectGroupBridgePairingRequest(requestId: string, rejectedBy = "", reason = "") {
  return apiJson<{ request: GroupBridgePairingRequest }>(
    `/api/group-bridge/pairing/requests/${encodeURIComponent(requestId)}/reject`,
    {
      method: "POST",
      body: JSON.stringify({ rejected_by: rejectedBy, reason }),
    },
  );
}

export async function fetchGroupBridgeTrusts(groupId?: string) {
  const suffix = groupId ? `?group_id=${encodeURIComponent(groupId)}` : "";
  return apiJson<{ trusts: GroupBridgeTrust[] }>(`/api/group-bridge/pairing/trusts${suffix}`);
}

export async function revokeGroupBridgeTrust(trustId: string, revokedBy = "") {
  return apiJson<{ trust: GroupBridgeTrust }>(
    `/api/group-bridge/pairing/trusts/${encodeURIComponent(trustId)}/revoke`,
    {
      method: "POST",
      body: JSON.stringify({ revoked_by: revokedBy }),
    },
  );
}

export async function updateGroupBridgeTrustAccess(trustId: string, accessLevel: GroupBridgeAccessLevel, updatedBy = "") {
  return apiJson<{ trust: GroupBridgeTrust }>(
    `/api/group-bridge/pairing/trusts/${encodeURIComponent(trustId)}/access`,
    {
      method: "POST",
      body: JSON.stringify({ access_level: accessLevel, updated_by: updatedBy }),
    },
  );
}

export async function refreshGroupBridgeTrustRemoteInfo(trustId: string) {
  return apiJson<{ trust: GroupBridgeTrust; remote_status: Record<string, unknown> }>(
    `/api/group-bridge/pairing/trusts/${encodeURIComponent(trustId)}/refresh`,
    { method: "POST" },
  );
}

export async function fetchGroupBridgePairingOutbounds(groupId?: string) {
  const suffix = groupId ? `?group_id=${encodeURIComponent(groupId)}` : "";
  return apiJson<{ outbounds: GroupBridgePairingOutbound[] }>(`/api/group-bridge/pairing/outbounds${suffix}`);
}

export async function syncGroupBridgePairingOutbound(outboundId: string) {
  return apiJson<{ outbound: GroupBridgePairingOutbound }>(
    `/api/group-bridge/pairing/outbounds/${encodeURIComponent(outboundId)}/sync`,
    { method: "POST" },
  );
}

export async function deleteGroupBridgePairingOutbound(outboundId: string) {
  return apiJson<{ deleted: boolean }>(
    `/api/group-bridge/pairing/outbounds/${encodeURIComponent(outboundId)}/delete`,
    { method: "POST" },
  );
}

File: src/cccc/ports/web/routes/group_bridge.py
"""Web routes for Group Bridge remote-send management (Stage 3).

Thin layer only: request parsing, principal/group authorization (re-checked
backend-side), delegation to the group_bridge kernel store / Stage 2 daemon op,
and secret-free response projection. No business logic, no transport I/O.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Depends, HTTPException, Request, WebSocket
from pydantic import BaseModel, ConfigDict, Field

from ....daemon.group_bridge.ops import try_handle_remote_send_op
from ....daemon.group_bridge.ws_endpoint import handle_group_bridge_session_websocket
from ....daemon.group_bridge.ws_session import send_via_session
from ....kernel.group_bridge.receipts import safe_error_projection
from ....kernel.group_bridge.pairing import (
    approve_pairing_request,
    create_pairing_invite,
    create_pairing_request,
    delete_pairing_outbound,
    get_pairing_invite,
    get_pairing_invite_for_code,
    get_pairing_request_public_status,
    get_local_identity,
    get_pairing_outbound,
    list_pairing_outbounds,
    list_pairing_requests,
    list_trusts,
    prepare_pairing_invite_for_connection_info,
    reject_pairing_request,
    revoke_trust,
    update_trust_access_level,
    update_trust_remote_info,
    upsert_pairing_outbound,
)
from ....kernel.group_bridge.pairing_remote import build_connection_payload, sync_remote_pairing_outbound, submit_remote_pairing_request
from ....kernel.group_bridge.registration import delete_registration, get_registration, list_registrations
from ....ports.mcp.common import MCPError
from ....ports.mcp.handlers.group_bridge_client import remote_access as group_bridge_remote_access
from ..schemas import RouteContext, check_group, get_principal, require_user

# Registration records may carry a credential reference. Public projections must
# never echo it back to UI/API callers.
_REGISTRATION_PUBLIC_FIELDS = (
    "registration_id",
    "group_id",
    "url",
    "transport",
    "remote_group_id",
    "remote_peer_id",
    "multiaddrs",
    "user_id",
    "status",
    "created_at",
    "updated_at",
    "last_sync_at",
    "last_error",
)


class GroupBridgeUnregisterRequest(BaseModel):
    registration_id: str


class PairingInviteCreateRequest(BaseModel):
    group_id: str
    remote_group_id: str = ""
    remote_peer_id: str = ""
    multiaddrs: List[str] = Field(default_factory=list)
    ttl_seconds: int = 600


class PairingConnectionInfoRequest(BaseModel):
    group_id: str
    invite_id: str
    issuer_endpoint: str
    issuer_group_title: str = ""


class PairingRequestCreateRequest(BaseModel):
    pairing_code: str
    requester_group_id: str
    requester_peer_id: str
    requester_endpoint: str = ""
    requester_multiaddrs: List[str] = Field(default_factory=list)
    invite_id: str = ""


class RemotePairingRequestCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    pairing_code: str = Field(min_length=1, max_length=64)
    invite_id: str = Field(min_length=1, max_length=128)
    requester_group_id: str = Field(min_length=1, max_length=256)
    requester_group_title: str = Field(default="", max_length=256)
    requester_endpoint: str = Field(default="", max_length=2048)
    requester_peer_id: str = Field(min_length=1, max_length=512)
    requester_node_id: str = Field(default="", max_length=512)
    requester_multiaddrs: List[str] = Field(default_factory=list, max_length=32)


class GroupBridgeSessionSendRequest(BaseModel):
    target_group_id: str
    src_group_id: str
    remote_peer_id: str
    request: Dict[str, Any] = Field(default_factory=dict)
    timeout: float = 5.0


class RemotePairingSubmitRequest(BaseModel):
    payload: Dict[str, Any]
    local_group_id: str
    local_group_title: str = ""


class PairingApproveRequest(BaseModel):
    approver_user_id: str = ""


class PairingRejectRequest(BaseModel):
    rejected_by: str = ""
    reason: str = ""


class TrustRevokeRequest(BaseModel):
    revoked_by: str = ""


class TrustAccessUpdateRequest(BaseModel):
    access_level: str
    updated_by: str = ""


def _project_registration(record: Dict[str, Any]) -> Dict[str, Any]:
    return {field: record.get(field) for field in _REGISTRATION_PUBLIC_FIELDS if field in record}


def _project_receipt(receipt: Optional[Dict[str, Any]]) -> Optional[Dict[str, Any]]:
    if not isinstance(receipt, dict):
        return None
    out = dict(receipt)
    if isinstance(out.get("error"), dict):
        out["error"] = safe_error_projection(out["error"])
    return out


def _require_group_or_403(request: Request, group_id: str) -> None:
    gid = str(group_id or "").strip()
    if not gid:
        raise HTTPException(status_code=400, detail={"code": "missing_group_id", "message": "group_id is required", "details": {}})
    # Re-check group authorization backend-side (admin-all vs explicit allow-list).
    check_group(request, gid)


def _principal_can_access(principal: Any, group_id: str) -> bool:
    if bool(getattr(principal, "is_admin", False)):
        return True
    allowed = getattr(principal, "allowed_groups", ()) or ()
    return str(group_id or "").strip() in {str(g or "").strip() for g in allowed}


def _filter_group_scoped_items(request: Request, items: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    principal = get_principal(request)
    return [item for item in items if _principal_can_access(principal, str(item.get("group_id") or ""))]


def _filter_local_group_scoped_items(request: Request, items: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    principal = get_principal(request)
    return [item for item in items if _principal_can_access(principal, str(item.get("local_group_id") or ""))]


def _require_loopback_client(request: Request) -> None:
    client = getattr(request, "client", None)
    host = str(getattr(client, "host", "") or "").strip().lower()
    if host not in {"127.0.0.1", "::1", "localhost"}:
        raise HTTPException(status_code=403, detail={"code": "forbidden", "message": "local Group Bridge session owner calls only", "details": {}})


def create_routers(ctx: RouteContext) -> list[APIRouter]:
    _ = ctx
    router = APIRouter(prefix="/api/group-bridge", dependencies=[Depends(require_user)])

    @router.get("/status")
    async def group_bridge_status(request: Request, group_id: str = "") -> Dict[str, Any]:
        gid = str(group_id or "").strip()
        if gid:
            _require_group_or_403(request, gid)
            items = [r for r in list_registrations() if str(r.get("group_id") or "") == gid]
        else:
            principal = get_principal(request)
            items = [r for r in list_registrations() if _principal_can_access(principal, str(r.get("group_id") or ""))]
        return {"ok": True, "result": {"registrations": [_project_registration(r) for r in items]}}

    @router.post("/unregister")
    async def group_bridge_unregister(request: Request, req: GroupBridgeUnregisterRequest) -> Dict[str, Any]:
        rid = str(req.registration_id or "").strip()
        record = get_registration(rid)
        if not record:
            raise HTTPException(status_code=404, detail={"code": "not_found", "message": "registration not found", "details": {}})
        _require_group_or_403(request, str(record.get("group_id") or ""))
        deleted = delete_registration(rid)
        return {"ok": True, "result": {"deleted": bool(deleted)}}

    @router.get("/registrations/{registration_id}/deliveries/{idempotency_key}")
    async def group_bridge_delivery_status(request: Request, registration_id: str, idempotency_key: str) -> Dict[str, Any]:
        rid = str(registration_id or "").strip()
        record = get_registration(rid)
        if not record:
            raise HTTPException(status_code=404, detail={"code": "not_found", "message": "registration not found", "details": {}})
        # Web-layer principal authz...
        _require_group_or_403(request, str(record.get("group_id") or ""))
        # ...and reuse the Stage 2 daemon op, which re-enforces the registration
        # group scope before returning the receipt.
        resp = try_handle_remote_send_op(
            "remote_delivery_status",
            {
                "group_id": str(record.get("group_id") or ""),
                "registration_id": rid,
                "idempotency_key": str(idempotency_key or "").strip(),
            },
        )
        if resp is None:
            raise HTTPException(status_code=500, detail={"code": "internal_error", "message": "delivery status unavailable", "details": {}})
        if not resp.ok:
            err = resp.error
            raise HTTPException(
                status_code=400,
                detail={"code": getattr(err, "code", "error"), "message": getattr(err, "message", ""), "details": getattr(err, "details", {}) or {}},
            )
        receipt = (resp.result or {}).get("receipt") if isinstance(resp.result, dict) else None
        return {"ok": True, "result": {"receipt": _project_receipt(receipt)}}

    @router.get("/pairing/identity")
    async def group_bridge_pairing_identity() -> Dict[str, Any]:
        return {"ok": True, "result": {"identity": get_local_identity()}}

    @router.post("/pairing/invites")
    async def group_bridge_pairing_invite_create(request: Request, req: PairingInviteCreateRequest) -> Dict[str, Any]:
        _require_group_or_403(request, req.group_id)
        try:
            invite = create_pairing_invite(
                group_id=req.group_id,
                remote_group_id=req.remote_group_id,
                remote_peer_id=req.remote_peer_id,
                multiaddrs=req.multiaddrs,
                ttl_seconds=req.ttl_seconds,
            )
        except ValueError as exc:
            raise HTTPException(status_code=400, detail={"code": "invalid_request", "message": str(exc), "details": {}}) from exc
        return {"ok": True, "result": {"invite": invite}}

    @router.post("/pairing/connection-info")
    async def group_bridge_pairing_connection_info(request: Request, req: PairingConnectionInfoRequest) -> Dict[str, Any]:
        gid = str(req.group_id or "").strip()
        _require_group_or_403(request, gid)
        invite = get_pairing_invite(req.invite_id)
        if not invite:
            raise HTTPException(status_code=404, detail={"code": "not_found", "message": "pairing invite not found", "details": {}})
        if str(invite.get("group_id") or "").strip() != gid:
            raise HTTPException(status_code=403, detail={"code": "forbidden", "message": "pairing invite is not in this group", "details": {}})
        try:
            invite = prepare_pairing_invite_for_connection_info(req.invite_id)
            payload = build_connection_payload(
                invite,
                issuer_endpoint=req.issuer_endpoint,
                issuer_group_title=req.issuer_group_title,
            )
        except ValueError as exc:
            raise HTTPException(status_code=400, detail={"code": "invalid_request", "message": str(exc), "details": {}}) from exc
        return {"ok": True, "result": {"payload": payload}}

    @router.post("/pairing/requests")
    async def group_bridge_pairing_request_create(request: Request, req: PairingRequestCreateRequest) -> Dict[str, Any]:
        _require_group_or_403(request, req.requester_group_id)
        invite = get_pairing_invite_for_code(req.pairing_code, invite_id=req.invite_id)
        if invite is not None:
            _require_group_or_403(request, str(invite.get("group_id") or ""))
        try:
            pairing_request = create_pairing_request(
                req.pairing_code,
                requester_group_id=req.requester_group_id,
                requester_group_title="",
                requester_peer_id=req.requester_peer_id,
                requester_endpoint=req.requester_endpoint,
                requester_multiaddrs=req.requester_multiaddrs,
                invite_id=req.invite_id,
            )
        except ValueError as exc:
            raise HTTPException(status_code=400, detail={"code": "invalid_request", "message": str(exc), "details": {}}) from exc
        return {"ok": True, "result": {"request": pairing_request}}

    @router.post("/pairing/remote-requests")
    async def group_bridge_pairing_remote_submit(request: Request, req: RemotePairingSubmitRequest) -> Dict[str, Any]:
        _require_group_or_403(request, req.local_group_id)
        try:
            outbound = submit_remote_pairing_request(
                req.payload,
                local_group_id=req.local_group_id,
                local_group_title=req.local_group_title,
                allow_localhost=True,
            )
        except ValueError as exc:
            raise HTTPException(status_code=400, detail={"code": "invalid_request", "message": str(exc), "details": {}}) from exc
        outbound = {**outbound, "local_group_id": str(outbound.get("local_group_id") or req.local_group_id)}
        outbound = upsert_pairing_outbound(outbound)
        return {"ok": True, "result": {"outbound": outbound}}

    @router.get("/pairing/requests")
    async def group_bridge_pairing_request_list(request: Request, group_id: str = "") -> Dict[str, Any]:
        gid = str(group_id or "").strip()
        if gid:
            _require_group_or_403(request, gid)
            items = list_pairing_requests(group_id=gid)
        else:
            items = _filter_group_scoped_items(request, list_pairing_requests())
        return {"ok": True, "result": {"requests": items}}

    @router.post("/pairing/requests/{request_id}/approve")
    async def group_bridge_pairing_approve(request: Request, request_id: str, req: PairingApproveRequest) -> Dict[str, Any]:
        existing = list_pairing_requests()
        match = next((item for item in existing if str(item.get("request_id") or "") == str(request_id or "").strip()), None)
        if not match:
            raise HTTPException(status_code=404, detail={"code": "not_found", "message": "pairing request not found", "details": {}})
        _require_group_or_403(request, str(match.get("group_id") or ""))
        try:
            approved = approve_pairing_request(request_id, approver_user_id=req.approver_user_id)
        except ValueError as exc:
            raise HTTPException(status_code=400, detail={"code": "invalid_request", "message": str(exc), "details": {}}) from exc
        return {
            "ok": True,
            "result": {
                "request": approved.get("request"),
                "registration": _project_registration(approved.get("registration") or {}),
                "trust": approved.get("trust"),
            },
        }

    @router.post("/pairing/requests/{request_id}/reject")
    async def group_bridge_pairing_reject(request: Request, request_id: str, req: PairingRejectRequest) -> Dict[str, Any]:
        existing = list_pairing_requests()
        match = next((item for item in existing if str(item.get("request_id") or "") == str(request_id or "").strip()), None)
        if not match:
            raise HTTPException(status_code=404, detail={"code": "not_found", "message": "pairing request not found", "details": {}})
        _require_group_or_403(request, str(match.get("group_id") or ""))
        try:
            rejected = reject_pairing_request(request_id, rejected_by=req.rejected_by, reason=req.reason)
        except ValueError as exc:
            raise HTTPException(status_code=400, detail={"code": "invalid_request", "message": str(exc), "details": {}}) from exc
        return {"ok": True, "result": {"request": rejected}}

    @router.get("/pairing/trusts")
    async def group_bridge_pairing_trust_list(request: Request, group_id: str = "") -> Dict[str, Any]:
        gid = str(group_id or "").strip()
        if gid:
            _require_group_or_403(request, gid)
            items = list_trusts(group_id=gid)
        else:
            items = _filter_group_scoped_items(request, list_trusts())
        return {"ok": True, "result": {"trusts": items}}

    @router.post("/pairing/trusts/{trust_id}/revoke")
    async def group_bridge_pairing_trust_revoke(request: Request, trust_id: str, req: TrustRevokeRequest) -> Dict[str, Any]:
        existing = list_trusts()
        match = next((item for item in existing if str(item.get("trust_id") or "") == str(trust_id or "").strip()), None)
        if not match:
            raise HTTPException(status_code=404, detail={"code": "not_found", "message": "trust not found", "details": {}})
        _require_group_or_403(request, str(match.get("group_id") or ""))
        try:
            revoked = revoke_trust(trust_id, revoked_by=req.revoked_by)
        except ValueError as exc:
            raise HTTPException(status_code=400, detail={"code": "invalid_request", "message": str(exc), "details": {}}) from exc
        return {"ok": True, "result": {"trust": revoked}}

    @router.post("/pairing/trusts/{trust_id}/access")
    async def group_bridge_pairing_trust_access_update(request: Request, trust_id: str, req: TrustAccessUpdateRequest) -> Dict[str, Any]:
        existing = list_trusts()
        match = next((item for item in existing if str(item.get("trust_id") or "") == str(trust_id or "").strip()), None)
        if not match:
            raise HTTPException(status_code=404, detail={"code": "not_found", "message": "trust not found", "details": {}})
        _require_group_or_403(request, str(match.get("group_id") or ""))
        try:
            updated = update_trust_access_level(trust_id, req.access_level, updated_by=req.updated_by)
        except ValueError as exc:
            raise HTTPException(status_code=400, detail={"code": "invalid_request", "message": str(exc), "details": {}}) from exc
        return {"ok": True, "result": {"trust": updated}}

    @router.post("/pairing/trusts/{trust_id}/refresh")
    async def group_bridge_pairing_trust_refresh(request: Request, trust_id: str) -> Dict[str, Any]:
        existing = list_trusts()
        match = next((item for item in existing if str(item.get("trust_id") or "") == str(trust_id or "").strip()), None)
        if not match:
            raise HTTPException(status_code=404, detail={"code": "not_found", "message": "trust not found", "details": {}})
        group_id = str(match.get("group_id") or "").strip()
        remote_group_id = str(match.get("remote_group_id") or "").strip()
        _require_group_or_403(request, group_id)
        try:
            remote_status = group_bridge_remote_access(
                group_id=group_id,
                arguments={"action": "status", "remote_group_id": remote_group_id},
            )
            updated = update_trust_remote_info(
                trust_id,
                remote_group_title=str(remote_status.get("remote_group_title") or ""),
                remote_access_level=str(remote_status.get("access_level") or ""),
            )
        except MCPError as exc:
            raise HTTPException(
                status_code=400,
                detail={"code": exc.code, "message": exc.message, "details": exc.details or {}},
            ) from exc
        except ValueError as exc:
            raise HTTPException(status_code=400, detail={"code": "invalid_request", "message": str(exc), "details": {}}) from exc
        return {"ok": True, "result": {"trust": updated, "remote_status": remote_status}}

    @router.get("/pairing/outbounds")
    async def group_bridge_pairing_outbound_list(request: Request, group_id: str = "") -> Dict[str, Any]:
        gid = str(group_id or "").strip()
        if gid:
            _require_group_or_403(request, gid)
            items = list_pairing_outbounds(group_id=gid)
        else:
            items = _filter_local_group_scoped_items(request, list_pairing_outbounds())
        return {"ok": True, "result": {"outbounds": items}}

    @router.post("/pairing/outbounds/{outbound_id}/sync")
    async def group_bridge_pairing_outbound_sync(request: Request, outbound_id: str) -> Dict[str, Any]:
        existing = get_pairing_outbound(outbound_id)
        if not existing:
            raise HTTPException(status_code=404, detail={"code": "not_found", "message": "pairing outbound not found", "details": {}})
        _require_group_or_403(request, str(existing.get("local_group_id") or ""))
        try:
            outbound = sync_remote_pairing_outbound(outbound_id, allow_localhost=True)
        except ValueError as exc:
            raise HTTPException(status_code=400, detail={"code": "invalid_request", "message": str(exc), "details": {}}) from exc
        if isinstance(outbound, dict) and isinstance(outbound.get("outbound"), dict):
            outbound = outbound["outbound"]
        return {"ok": True, "result": {"outbound": outbound}}

    @router.post("/pairing/outbounds/{outbound_id}/delete")
    async def group_bridge_pairing_outbound_delete(request: Request, outbound_id: str) -> Dict[str, Any]:
        existing = get_pairing_outbound(outbound_id)
        if not existing:
            raise HTTPException(status_code=404, detail={"code": "not_found", "message": "pairing outbound not found", "details": {}})
        _require_group_or_403(request, str(existing.get("local_group_id") or ""))
        return {"ok": True, "result": {"deleted": delete_pairing_outbound(outbound_id)}}

    public_router = APIRouter(prefix="/api/group-bridge")

    @public_router.post("/pairing/requests/remote")
    async def group_bridge_pairing_remote_request_create(req: RemotePairingRequestCreateRequest) -> Dict[str, Any]:
        try:
            pairing_request = create_pairing_request(
                req.pairing_code,
                requester_group_id=req.requester_group_id,
                requester_group_title=req.requester_group_title,
                requester_peer_id=req.requester_peer_id,
                requester_endpoint=req.requester_endpoint,
                requester_multiaddrs=req.requester_multiaddrs,
                invite_id=req.invite_id,
            )
        except ValueError as exc:
            raise HTTPException(status_code=400, detail={"code": "invalid_request", "message": str(exc), "details": {}}) from exc
        return {"ok": True, "result": {"request": pairing_request}}

    @public_router.get("/pairing/requests/remote/status")
    async def group_bridge_pairing_remote_request_status(request_id: str = "", invite_id: str = "") -> Dict[str, Any]:
        pairing_request = get_pairing_request_public_status(request_id, invite_id=invite_id)
        if not pairing_request:
            raise HTTPException(status_code=404, detail={"code": "not_found", "message": "pairing request not found", "details": {}})
        return {"ok": True, "result": {"request": pairing_request}}

    @public_router.websocket("/session/ws")
    async def group_bridge_session_ws(websocket: WebSocket) -> None:
        await handle_group_bridge_session_websocket(websocket)

    @public_router.post("/session/send")
    async def group_bridge_session_send(request: Request, req: GroupBridgeSessionSendRequest) -> Dict[str, Any]:
        _require_loopback_client(request)
        return await send_via_session(
            target_group_id=req.target_group_id,
            src_group_id=req.src_group_id,
            remote_peer_id=req.remote_peer_id,
            request=req.request,
            timeout=req.timeout,
        )

    return [public_router, router]
